### ISSUE TITLE
Compile error: message.h file not found

### ISSUE THREAD

**Issue by shinchiro** [KhronosGroup/glslang#1477]
SPIRV-Tools remove the 'message.h' file recently
https://github.com/KhronosGroup/SPIRV-Tools/pull/1838

So the compile failed with "SPIRV/GlslangToSpv.cpp:59:14: fatal error: message.h: No such file or directory"

**Comment by johnkslang** [comment]
I'm guessing that the compile fails with top of tree, versus with the known good commit brought in by `glslang/update_glslang_sources.py`. Yes?

This may be a "LunarG SDK" (they currently suggest the known good) versus "keep top of trees working" issue. I prefer the latter, or possibly a more community driven choice, but can see trade-offs.

**Comment by LaserEyess** [comment]
I'm also getting the same issue, fwiw. @johnkslang, are you saying there is some branch/commit I should be checking out to compile this properly? I am using [this fork](https://github.com/google/glslang) and compiling with [these instructions (1 and then 4a)](https://github.com/google/shaderc#getting-and-building-shaderc) but I am unsure if that matters. I tested with this repo as well and it doesn't seem to make a difference. GhslangToSpv.cpp still has a reference to `StringifyMessage` and it wants `message.h` for that.

**Comment by johnkslang** [comment]
The glslang in this repository is currently tested with the SPIR-V tools in `glslang/update_glslang_sources.py`.

For this repository, I could see stop having `glslang/update_glslang_sources.py` being a LunarG chosen commit and move to something closer to top of tree, but would like more input on that.

Having upstream components delete files certainly makes things more difficult, but I think that needs be dealt with in https://github.com/KhronosGroup/SPIRV-Tools/pull/1838.

**Comment by dneto0** [comment]
> Having upstream components delete files certainly makes things more difficult, but I think that needs be dealt with in KhronosGroup/SPIRV-Tools#1838.

That message.h and message.cpp were internal implementation details. It was not in the public API of SPIRV-Tools. (The public API is in SPIRV-Tools'   include/ directory.)

I recommend GlslangToSpirv.cpp implement its own stringifier.

**Comment by johnkslang** [comment]
> That message.h and message.cpp were internal implementation details. It was not in the public API of SPIRV-Tools. (The public API is in SPIRV-Tools' include/ directory.)

I see. I didn't realize, as the  inclusion of that came from someone (@greg-lunarg?) working on the tools, so I thought it was the correct way.

> I recommend GlslangToSpirv.cpp implement its own stringifier.

I see. I'm a bit surprised to need one, since A) it already has something very much like that for its messages, and B) I thought it was just for the optimizer which should be silent unless the SPIR-V is broken (basically an internal error, not a message back to the user).

Thanks for the substitute, but it seems to also point to a bigger question about why this is necessary.  Should the stringifier thing be more of an "Internal ERROR:..." thing when it is part of Glslang with code that has already been accepted with no errors by the front end?

**Comment by johnkslang** [comment]
I see.  More typically, I thought there is a need to get a message back if bad SPIR-V was generated by the front end.

The point is these are exceptional situations, where (in practice) it is outside the user's control, rather than things like "info", "warning", "error" as if it is code the user can act on.

I see now how the suggestion maps internal and fatal errors to the "error:" (glslang proper uses "INTERNAL ERROR" for those cases).

Would it make sense for glslang's stringifier for the optimizer to map some of these to "INTERNAL ERROR"? Or, less extreme, to mimic glslang's existing form of message delivery, and/or include "optimizer" in it?  "OPTIMIZING ERROR:"  Having thought through this, it does seem like the value in parameterizing the stringifier, so I should probably do something like that. Better suggestions?

**Comment by antiagainst** [comment]
Please don't hate Lei for the large `spv_message_level_t` enum. ;-P

IMHO, from a compiler's point of view, we should try to be as verbose as possible when something goes wrong. Additional information provided to developers on errors is always useful.

**Comment by johnkslang** [comment]
Yes, I'll do the upgrade. Thanks for making the design point clear.

@antiagainst agreed. What is interesting in this case is that if the developer is making a mistake, the front end should complain, nothing else, so it's almost certainly a compiler bug, either in the front end, or AST -> SPIR-V translator, or the optimizer.  E.g.,

```
SPV_MSG_ERROR,           // Normal error due to user input.
```

That comment is correct if the SPIR-V came from that user. However, I suspect it means:

```
SPV_MSG_ERROR,           // Error due to incorrect SPIR-V semantics
```

Or, `syntax or semantics`, or whatever triggers that error, and then glslang should map it to an internal error, because it's its own fault that happened.

So, it is nice we have a chance to remap how the errors are reported.

**Comment by arcady-lunarg** [comment]
This was fixed in commit 9fd2a8cdfc15883059500c243949dd89dc19d621.
